### ISSUE TITLE
Running gradio app using script from colab

### ISSUE THREAD

**Issue by innat** [gradio-app/gradio#4403]
### Describe the bug

When trying to run gradio app from colab, the app doesn't run from a **written source code file** but does run if we place the source code in cell. 

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

1. Placing the following codes in colab cell, the program runs as it should be.

````python
import gradio as gr
import torch
import requests
from torchvision import transforms

model = torch.hub.load('pytorch/vision:v0.6.0', 'resnet18', pretrained=True).eval()
response = requests.get("https://git.io/JJkYN")
labels = response.text.split("\n")

def predict(inp):
  inp = transforms.ToTensor()(inp).unsqueeze(0)
  with torch.no_grad():
    prediction = torch.nn.functional.softmax(model(inp)[0], dim=0)
    confidences = {labels[i]: float(prediction[i]) for i in range(1000)}    
  return confidences

demo = gr.Interface(
    fn=predict, 
    inputs=gr.inputs.Image(type="pil"),
    outputs=gr.outputs.Label(num_top_classes=3),
)
demo.launch()
````

2. If I place the above code in a script, let's say `app.py`, it just keep showing the following output progress and the host address wouldn't live.

![image](https://github.com/gradio-app/gradio/assets/17668390/dac87fef-648e-4e87-9d34-c56dea465e95)

However, if I add `demo.launch(share=True)` to the `app.py` script, then it gives 

Running on local URL:  http://127.0.0.1:7860/
Running on public URL: https://a406b46d16e8060ee5.gradio.live/

Among which the pulic URL becomes live for 72 hrs but the local URL stills off.

### Logs

```shell
Did you `debug=True`, nothing usefull shows up.
```

### System Info

```shell
System: Colab. 
Gradio: 3.33.1
```

### Severity

annoying

**Comment by abidlabs** [comment]
Hi @innat this is actually the expected behavior. The problem is that when running code from a script in Colab, you are actually running on a remote jupyter notebook, which means when the server spins up on port 7860, it won't be accessible to you on your local machine. It is actually running on _Colab_'s port 7860. Normally, you'd just SSH into the remote machine and forward the port to your local machine, but as far as I know, there's no way to SSH into a Colab notebook machine.

When you run code in a regular Colab cell, Gradio is able to detect that and override the default behavior of Colab so that when you click on the localhost link, it actually opens up a remote link, but we can't detect that if you are running your code via a script. 

So in short, really the best solution if you are running on Colab is to set `share=True`, as this will always work. Yes, there is a 72 hour timeout, but this is probably longer than your Colab session anyways, so it should hopefully suffice for your needs.
